File: Example_StaticLibrary/Demo/Growth/GenericShareViewController.m
/*
 * Copyright (c) StreetHawk, All rights reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

#import "GenericShareViewController.h"
#import <MessageUI/MessageUI.h>

@interface GenericShareViewController () <MFMailComposeViewControllerDelegate>

@property (nonatomic, strong) NSArray *arrayCells;

@end

@implementation GenericShareViewController

#pragma mark - life cycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.arrayCells = @[self.cellID, self.cellSource, self.cellMedium, self.cellContent, self.cellTerm, self.cellUrl, self.cellDestinationUrl, self.cellEmailSubject, self.cellEmailBody, self.cellShare];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.arrayCells.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = self.arrayCells[indexPath.row];
    return cell.bounds.size.height;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return self.arrayCells[indexPath.row];;
}

#pragma mark - UITextFieldDelegate handler

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [textField resignFirstResponder];
    return YES;
}

#pragma mark - event handler

- (IBAction)buttonShareClicked:(id)sender
{
    NSURL *deeplinkingUrl = nil;
    if (self.textboxUrl.text.length > 0)
    {
        deeplinkingUrl = [NSURL URLWithString:self.textboxUrl.text];
        if (deeplinkingUrl == nil)
        {
            UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"Deeplinking url format is invalid. Correct it or delete it." message:nil delegate:nil cancelButtonTitle:nil otherButtonTitles:@"OK", nil];
            [alertView show];
            return;
        }
    }
    NSURL *destinationUrl = nil;
    if (self.textboxDestinationUrl.text.length > 0)
    {
        destinationUrl = [NSURL URLWithString:self.textboxDestinationUrl.text];
        if (destinationUrl == nil)
        {
            UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"Destination url format is invalid. Correct it or delete it." message:nil delegate:nil cancelButtonTitle:nil otherButtonTitles:@"OK", nil];
            [alertView show];
            return;
        }
    }
    [StreetHawk originateShareWithCampaign:self.textboxID.text withSource:self.textboxSource.text withMedium:self.textboxMedium.text withContent:self.textboxContent.text withTerm:self.textboxTerm.text shareUrl:deeplinkingUrl withDefaultUrl:destinationUrl streetHawkGrowth_object:^(NSObject *result, NSError *error)
    {
        shPresentErrorAlertOrLog(error);
        if (error == nil)
        {
            dispatch_async(dispatch_get_main_queue(), ^
               {
                   NSString *shareUrl = (NSString *)result;
                   if ([MFMailComposeViewController canSendMail])
                   {
                       MFMailComposeViewController *mc = [[MFMailComposeViewController alloc] init];
                       mc.mailComposeDelegate = self;
                       [mc setMessageBody:[NSString stringWithFormat:@"%@\n\n%@", self.textboxEmailBody.text, shareUrl] isHTML:NO];
                       [mc setSubject:self.textboxEmailSubject.text];
                       [self presentViewController:mc animated:YES completion:nil];
                   }
                   else
                   {
                       UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"share_guid_url" message:shareUrl delegate:nil cancelButtonTitle:nil otherButtonTitles:@"OK", nil];
                       [alert show];
                   }
               });
        }
    }];
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
    if (!error)
    {
        [controller dismissViewControllerAnimated:YES completion:nil];
    }
}

@end

File: Example_Swift/Demo/Growth/GenericShareViewController.swift
/*
 * Copyright (c) StreetHawk, All rights reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

import UIKit
fileprivate func < <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l < r
  case (nil, _?):
    return true
  default:
    return false
  }
}

fileprivate func > <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l > r
  default:
    return rhs < lhs
  }
}


class GenericShareViewController: StreetHawkBaseTableViewController, UITextFieldDelegate
{
    @IBOutlet var cellID: UITableViewCell!
    @IBOutlet var textboxID: UITextField!
    @IBOutlet var cellSource: UITableViewCell!
    @IBOutlet var textboxSource: UITextField!
    @IBOutlet var cellMedium: UITableViewCell!
    @IBOutlet var textboxMedium: UITextField!
    @IBOutlet var cellContent: UITableViewCell!
    @IBOutlet var textboxContent: UITextField!
    @IBOutlet var cellTerm: UITableViewCell!
    @IBOutlet var textboxTerm: UITextField!
    @IBOutlet var cellUrl: UITableViewCell!
    @IBOutlet var textboxUrl: UITextField!
    @IBOutlet var cellDestinationUrl: UITableViewCell!
    @IBOutlet var textboxDestinationUrl: UITextField!
    @IBOutlet var cellEmailSubject: UITableViewCell!
    @IBOutlet var textboxEmailSubject: UITextField!
    @IBOutlet var cellEmailBody: UITableViewCell!
    @IBOutlet var textboxEmailBody: UITextField!
    @IBOutlet var cellShare: UITableViewCell!
    
    var arrayCells : [UITableViewCell] = []
    
    //life cycle
    
    required init?(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder)
    }
    
    override init(style: UITableViewStyle)
    {
        super.init(style: .plain)
    }
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?)
    {
        super.init(nibName: "GenericShareViewController"/*must explict write name, use nil not load correct xib in iOS 7*/, bundle: nibBundleOrNil)
    }
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        self.arrayCells = [self.cellID, self.cellSource, self.cellMedium, self.cellContent, self.cellTerm, self.cellUrl, self.cellDestinationUrl, self.cellEmailSubject, self.cellEmailBody, self.cellShare]
    }
    
    //Table view data source
    
    override func numberOfSections(in tableView: UITableView) -> Int
    {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return self.arrayCells.count
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat
    {
        let cell = self.arrayCells[(indexPath as NSIndexPath).row]
        return cell.bounds.size.height
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        return self.arrayCells[(indexPath as NSIndexPath).row]
    }
    
    //UITextFieldDelegate handler
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool
    {
        textField.resignFirstResponder()
        return true
    }
    
    //event handler
    
    @IBAction func buttonShareClicked(_ sender: AnyObject)
    {
        var deeplinkingUrl : URL?
        if (self.textboxUrl.text?.lengthOfBytes(using: String.Encoding.utf8) > 0)
        {
            deeplinkingUrl = URL(string: self.textboxUrl.text!)
            if !(deeplinkingUrl != nil)
            {
                let alertView = UIAlertView(title: "Deeplinking url format is invalid. Correct it or delete it.", message: nil, delegate: nil, cancelButtonTitle: "OK")
                alertView.show()
                return
            }
        }
        var destinationUrl : URL?
        if (self.textboxDestinationUrl.text?.lengthOfBytes(using: String.Encoding.utf8) > 0)
        {
            destinationUrl = URL(string: self.textboxDestinationUrl.text!)
            if !(destinationUrl != nil)
            {
                let alertView = UIAlertView(title: "Destination url format is invalid. Correct it or delete it.", message: nil, delegate: nil, cancelButtonTitle: "OK")
                alertView.show()
                return
            }
        }
        SHApp.sharedInstance().originateShare(withCampaign: self.textboxID.text, withSource: self.textboxSource.text, withMedium: self.textboxMedium.text, withContent: self.textboxContent.text, withTerm: self.textboxTerm.text, share: deeplinkingUrl, withDefaultUrl: destinationUrl, streetHawkGrowth_object: { (result, error) in
            DispatchQueue.main.async(execute: {
                if !(error != nil)
                {
                    let shareUrl = result as! String
                    let alert = UIAlertView(title: "share_guid_url", message: shareUrl, delegate: nil, cancelButtonTitle: "OK")
                    alert.show()
                }
                else
                {
                    let alert = UIAlertView(title: "Error", message: error?.localizedDescription, delegate: nil, cancelButtonTitle: "OK")
                    alert.show()
                }
            })
        })
    }
}
